### PR TITLE
CTests C48_S2SW GFS Wave Expansion

### LINKED ISSUE #4318: Add 6 New CTest Cases for Atmosphere and C48 S2SW gfs Components
### **C48_S2SW Configuration (6 new tests)**

Coupled system tests to complement existing `gfs_fcst_seg0`, `gfs_ocean_prod_f006`, `gfs_ice_prod_f006`:

#### 4. **C48_S2SW-gfs_stage**
- **Purpose**: Validate staging of coupled model ICs (atmosphere + ocean + ice + wave)
- **Input**: 
  - 13 atmosphere IC files
  - 3 restart files (ocean, ice, wave from previous cycle)
  - 1 wave prep file
- **Output**: Complete staged coupled ICs in ROTDIR
- **Test Focus**:
  - Multi-component staging coordination
  - Restart file handling (6-hour offset pattern)
  - Component consistency checks
  - Missing component detection

#### 5. **C48_S2SW-gfs_waveinit**
- **Purpose**: Test wave model initialization
- **Input**: Wave prep file + staged ICs
- **Output**: Wave initialization files (mod_def.ww3, ww3_grid.inp)
- **Test Focus**:
  - WaveWatch III grid initialization
  - Boundary condition setup
  - Configuration file generation
  - Grid consistency validation

#### 6. **C48_S2SW-gfs_wavepostsbs_f000-f002**
- **Purpose**: Wave post-processing for spectral bin statistics (hours 0, 1, 2)
- **Input**: Wave forecast output (f000, f001, f002 from `gfs_fcst_seg0`)
- **Output**: Post-processed spectral statistics files
- **Test Focus**:
  - Spectral binning accuracy
  - Multi-hour processing
  - Output file completeness
  - Missing hour handling

#### 7. **C48_S2SW-gfs_wavebndpnt**
- **Purpose**: Wave boundary point extraction
- **Input**: Wave forecast output (f000-f048)
- **Output**: Boundary point data files
- **Test Focus**:
  - Boundary point identification
  - Data extraction accuracy
  - Time series construction
  - Coordinate consistency

#### 8. **C48_S2SW-gfs_wavebndpntbll**
- **Purpose**: Wave boundary point bulletin generation
- **Input**: Boundary point data from `gfs_wavebndpnt`
- **Output**: Formatted bulletins for dissemination
- **Test Focus**:
  - Bulletin formatting compliance
  - WMO header generation
  - Data validation
  - Missing data flags

#### 9. **C48_S2SW-gfs_wavepostpnt**
- **Purpose**: Wave point output post-processing
- **Input**: Wave forecast point output
- **Output**: Post-processed point data (NetCDF format)
- **Test Focus**:
  - Point output extraction
  - NetCDF file generation
  - Metadata completeness
  - Quality control flags

---

### Phase 1: C48_ATM Tests
1. Create YAML configurations following existing patterns:
   - `C48_ATM-gfs_stage_ic.yaml`
   - `C48_ATM-gfs_tracker.yaml`
   - `C48_ATM-gfs_genesis.yaml`
2. Stage input data from nightly baseline runs
3. Implement validation scripts for outputs
4. Integrate with CMake test framework

### Phase 2: C48_S2SW Tests
1. Create YAML configurations for 6 wave-related tests
2. Coordinate with wave model team for validation criteria
3. Stage coupled system inputs (multi-component dependencies)
4. Implement wave-specific validation (spectral data, boundary points)
5. Test multi-cycle staging patterns (restart file handling)

### Phase 3: Integration & Documentation
1. Update `dev/ctests/README.md` with new test descriptions
2. Add test cases to CI/CD pipeline
3. Document expected runtime and resource requirements
4. Create troubleshooting guide for common failures

---

### Input Staging
All tests will use input data from **stable nightly baseline runs** staged in `$STAGED_CTESTS`:
- **C48_ATM tests**: Stage from GFS atmosphere-only runs (12Z cycle)
- **C48_S2SW tests**: Stage from coupled system runs (12Z cycle with 06Z restarts)

### Storage Requirements
- **C48_ATM tests**: ~2 GB per test (ICs + forecast history)
- **C48_S2SW tests**: ~4 GB per test (coupled ICs + multi-component outputs)
- **Total estimated**: ~30 GB for all 9 tests

---

### For Each Test Case
- ✅ YAML configuration file follows naming convention: `CASE-JOB.yaml`
- ✅ Input files staged correctly from baseline runs
- ✅ Output files generated and validated against expected results
- ✅ Test runs successfully in isolated EXPDIR environment
- ✅ Pass/fail criteria clearly defined and documented
- ✅ Integration with CMake CTest framework complete
- ✅ CI/CD pipeline includes new tests

### Overall Project
- ✅ All 9 tests pass consistently on Hera, Hercules, Orion
- ✅ Runtime < 30 minutes per test (parallel execution)
- ✅ Documentation updated with test descriptions and usage
- ✅ Zero false positives in 10 consecutive CI runs

### PR DESCRIPTION
**Pull request by TerrenceMcGuinness-NOAA**
# CTest Framework Expansion<br>C48_S2SW Wave Component Tests

_Claud Opus 4.5 was used to assist with developing this PR
This code requires review and validation by NWS staff_

## Summary

This PR adds comprehensive CTest coverage for wave component jobs in the C48_S2SW workflow configuration. Five new test cases validate critical wave post-processing tasks, expanding the automated testing framework for the S2SW (Subseasonal-to-Seasonal with Waves) system.

Resolves #4318 

### New Test Cases

| Test | Job | Description |
|------|-----|-------------|
| `C48_S2SW-gfs_stage_ic` | `gfs_stage_ic` | Stage initial conditions - stages atmosphere, ocean, ice, and wave restart files for coupled forecast |
| `C48_S2SW-gfs_waveinit` | `gfs_waveinit` | Wave initialization - generates mod_def files for 8 wave grids |
| `C48_S2SW-gfs_wavepostpnt` | `gfs_wavepostpnt` | Wave point post-processing - creates spectral and bulletin tar archives |
| `C48_S2SW-gfs_wavepostsbs_f000-f002` | `gfs_wavepostsbs_f000-f002` | Wave gridded post-processing - generates GRIB2 output for first forecast segment |
| `C48_S2SW-gfs_wavepostbndpnt` | `gfs_wavepostbndpnt` | Wave boundary point post-processing - creates interpolated boundary point spectra |
| `C48_S2SW-gfs_wavepostbndpntbll` | `gfs_wavepostbndpntbll` | Wave boundary point bulletin generation - creates WMO bulletins for boundary points |

### Files Added
- `dev/ctests/cases/C48_S2SW-gfs_stage_ic.yaml`
- `dev/ctests/cases/C48_S2SW-gfs_waveinit.yaml`
- `dev/ctests/cases/C48_S2SW-gfs_wavepostpnt.yaml`
- `dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml`
- `dev/ctests/cases/C48_S2SW-gfs_wavepostbndpnt.yaml`
- `dev/ctests/cases/C48_S2SW-gfs_wavepostbndpntbll.yaml`

### Files Modified
- `dev/ctests/CMakeLists.txt` - Added 6 new `AddJJOBTest()` entries

### gfs_waveinit
- **Purpose**: Validates wave model initialization and mod_def file generation
- **Inputs**: None (generation job)
- **Outputs**: 8 mod_def binary files for wave grids:
  - `uglo_100km` (native unstructured global)
  - `glo_30m` (global 30-minute)
  - `aoc_9km` (Arctic Ocean 9km)
  - `at_10m`, `ep_10m`, `wc_10m` (regional 10-minute grids)
  - `gnh_10m`, `gsh_15m` (hemispheric grids)

### gfs_wavepostpnt
- **Purpose**: Validates wave point output post-processing
- **Inputs**: 121 point files (f000-f120), 8 mod_def files
- **Outputs**: `spec.tar.gz` (spectral data), `bull.tar` (bulletins)

### gfs_wavepostsbs_f000-f002
- **Purpose**: Validates wave gridded side-by-side post-processing (first segment of metatask)
- **Inputs**: 3 uglo_100km history files (f000-f002), 8 mod_def files
- **Outputs**: GRIB2 files for `global.0p16` and `global.0p50` grids

### gfs_wavepostbndpnt
- **Purpose**: Validates interpolated boundary point spectral output
- **Inputs**: 121 point .nc and .log files (f000-f120), 8 mod_def files
- **Outputs**: `ibp.tar` (interpolated boundary point spectra)

### gfs_wavepostbndpntbll
- **Purpose**: Validates WMO bulletin generation for boundary points
- **Inputs**: 121 point .nc and .log files (f000-f120), 8 mod_def files
- **Outputs**: `ibpbull.tar` (boundary point bulletins), `ibpcbull.tar` (combined bulletins)

## Testing

All 60 CTests (15 tests × 4 phases) validated successfully on Hercules HPC:

| Workflow | Tests | Test Names |
|----------|-------|------------|
| C48_ATM-gfs | 5 | `stage_ic`, `fcst_seg0`, `atmos_prod_f000-f002`, `tracker`, `genesis` |
| C48_S2SW-gfs | 9 | `stage_ic`, `fcst_seg0`, `ocean_prod_f006`, `ice_prod_f006`, `waveinit`, `wavepostpnt`, `wavepostsbs_f000-f002`, `wavepostbndpnt`, `wavepostbndpntbll` |
| C48_S2SWA_gefs | 1 | `fcst_mem001_seg0` |
| **Total** | **15** | **60 test phases (15 tests × 4 phases each)** |

```
100% tests passed, 0 tests failed out of 60

Label Time Summary:
C48_ATM-gfs_atmos_prod_f000-f002        = 187.73 sec*proc (4 tests)
C48_ATM-gfs_fcst_seg0                   = 1157.70 sec*proc (4 tests)
C48_ATM-gfs_genesis                     = 493.46 sec*proc (4 tests)
C48_ATM-gfs_stage_ic                    =  67.41 sec*proc (4 tests)
C48_ATM-gfs_tracker                     =  73.16 sec*proc (4 tests)
C48_S2SW-gfs_fcst_seg0                  = 1166.68 sec*proc (4 tests)
C48_S2SW-gfs_ice_prod_f006              =  67.42 sec*proc (4 tests)
C48_S2SW-gfs_ocean_prod_f006            =  67.41 sec*proc (4 tests)
C48_S2SW-gfs_stage_ic                   =  65.12 sec*proc (4 tests)
C48_S2SW-gfs_waveinit                   =  67.42 sec*proc (4 tests)
C48_S2SW-gfs_wavepostbndpnt             = 258.04 sec*proc (4 tests)
C48_S2SW-gfs_wavepostbndpntbll          =  77.78 sec*proc (4 tests)
C48_S2SW-gfs_wavepostpnt                = 137.94 sec*proc (4 tests)
C48_S2SW-gfs_wavepostsbs_f000-f002      =  67.58 sec*proc (4 tests)
C48_S2SWA_gefs-gefs_fcst_mem001_seg0    = 793.20 sec*proc (4 tests)

Total Test time (real) = 1166.69 sec
```
## Technical Notes

1. **Wave History File Pattern**: The `uglo_100km` history files follow a non-uniform pattern:
   - f000-f048: hourly (49 files)
   - f051-f120: every 3 hours (24 files)
   - Total: 73 files

2. **Wave Point File Pattern**: Point output files exist for all 121 hours (f000-f120)

3. **Metatask Handling**: The `wavepostsbs` task is a Rocoto metatask requiring specific subtask names (e.g., `gfs_wavepostsbs_f000-f002`) for `rocotoboot`

4. **Task Output Separation**: 
   - `wavepostbndpnt` produces only `ibp.tar`
   - `wavepostbndpntbll` (separate task) produces `ibpbull.tar` and `ibpcbull.tar`

## Baseline Reference

- **Experiment**: `C48_S2SW_b40eab88-7222`
- **Location**: `/scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/HERA/BUILDS/GITLAB/stable/RUNTESTS`

## Dependencies

- Requires Rocoto dryrun build at `GFS_CI_ROCOTO_PATH`
- Uses `fv3-cpu` allocation on Hera
- Baseline data must be present at `STAGED_CTESTS`

## Checklist

- [x] YAML test cases follow existing patterns
- [x] CMakeLists.txt entries use correct `AddJJOBTest()` format
- [x] All 4 test phases pass for each test case (6 tests total)
- [x] Input/output file paths match baseline structure
- [x] Jinja2 templating with wxflow filters works correctly

### CODE REVIEW

File: dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        - [{{ SRC_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.ep_10m.bin, {{ DST_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.ep_10m.bin]
+        - [{{ SRC_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.wc_10m.bin, {{ DST_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.wc_10m.bin]
+        - [{{ SRC_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.gnh_10m.bin, {{ DST_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.gnh_10m.bin]
+        - [{{ SRC_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.gsh_15m.bin, {{ DST_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/prep/gfs.t{{ cyc }}z.mod_def.gsh_15m.bin]
+
+        # Wave gridded binary output files from forecast (f000-f002 for this segment)
+{% for fhr in range(0, 3) %}
+        - [{{ SRC_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/history/gfs.t{{ cyc }}z.uglo_100km.f{{ '%03d'|format(fhr) }}.bin, {{ DST_DIR }}/gfs.{{ PDY }}/{{ cyc }}/model/wave/history/gfs.t{{ cyc }}z.uglo_100km.f{{ '%03d'|format(fhr) }}.bin]
+{% endfor %}
```
Comment (DavidHuber-NOAA) [inline dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml]: Should this be explicit rather than using a `for` loop?

File: dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+output_files:
+    cmpfiles:
+        # GRIB2 output files for global 0.16 degree grid (f000-f002)
+{% for fhr in range(0, 3) %}
+        - [{{ SRC_DIR }}/gfs.{{ PDY }}/{{ cyc }}/products/wave/gridded/global.0p16/gfs.t{{ cyc }}z.global.0p16.f{{ '%03d'|format(fhr) }}.grib2, {{ DST_DIR }}/gfs.{{ PDY }}/{{ cyc }}/products/wave/gridded/global.0p16/gfs.t{{ cyc }}z.global.0p16.f{{ '%03d'|format(fhr) }}.grib2]
+{% endfor %}
+        # GRIB2 output files for global 0.50 degree grid (f000-f002)
+{% for fhr in range(0, 3) %}
+        - [{{ SRC_DIR }}/gfs.{{ PDY }}/{{ cyc }}/products/wave/gridded/global.0p50/gfs.t{{ cyc }}z.global.0p50.f{{ '%03d'|format(fhr) }}.grib2, {{ DST_DIR }}/gfs.{{ PDY }}/{{ cyc }}/products/wave/gridded/global.0p50/gfs.t{{ cyc }}z.global.0p50.f{{ '%03d'|format(fhr) }}.grib2]
+{% endfor %}
```
Comment (DavidHuber-NOAA) [inline dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml]: Same Q.

File: dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml (hunk 1 above)
Comment (TerrenceMcGuinness-NOAA) [reply dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml]: Yes!  Thanks.  ~T.McG

File: dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml (hunk 2 above)
Comment (TerrenceMcGuinness-NOAA) [reply dev/ctests/cases/C48_S2SW-gfs_wavepostsbs_f000-f002.yaml]: Same, yes.  Done.

File: dev/ctests/CMakeLists.txt [hunk 3]
```diff
@@ -76,26 +76,31 @@ function(AddJJOBTest)
...
-  set_tests_properties(test_${TEST_NAME}_stage PROPERTIES DEPENDS test_${TEST_NAME}_setup LABELS "${ARG_CASE}-${ARG_JOB}")
+  set_tests_properties(test_${TEST_NAME}_stage PROPERTIES DEPENDS test_${TEST_NAME}_setup LABELS "${TEST_NAME}")
 
   add_test(NAME test_${TEST_NAME}_execute
       COMMAND ./execute.sh ${TEST_NAME} ${ARG_JOB} ${ARG_TEST_DATE}
       WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/scripts)
-  set_tests_properties(test_${TEST_NAME}_execute PROPERTIES DEPENDS test_${TEST_NAME}_stage LABELS "${ARG_CASE}-${ARG_JOB}")
+  set_tests_properties(test_${TEST_NAME}_execute PROPERTIES DEPENDS test_${TEST_NAME}_stage LABELS "${TEST_NAME}")
 
   # TODO - This is a stub for the validation step
```
Comment (DavidHuber-NOAA) [inline dev/ctests/CMakeLists.txt:99]: Is this still a `TODO`?

**Review by DavidHuber-NOAA — APPROVED** [review]
Looks good. Just one question.
